drivers/dai/intel/dmic/dmic_nhlt.c

```diff
@@ -16,25 +16,167 @@ LOG_MODULE_REGISTER(LOG_DOMAIN);
+{
+	if (!packed) {
+		coeff += length;
+	} else {
+		coeff += ROUND_UP(3 * length, sizeof(uint32_t)) / sizeof(uint32_t);
+	}
+
+	return coeff;
+}
+
+/*
+ * @brief Write the fir coefficients in the PDMs' RAM
+ */
+static void dai_dmic_write_coeff(const struct dai_intel_dmic *dmic, uint32_t base,
+				 const uint32_t *coeff, int length, const bool packed)
+{
+	const uint8_t *coeff_in_bytes;
+	uint32_t coeff_val;
+
+	if (!packed) {
+		while (length--) {
+			dai_dmic_write(dmic, base, *coeff++);
+			base += sizeof(uint32_t);
+		}
+	} else {
+		coeff_in_bytes = (const uint8_t *)coeff;
+
+		while (length--) {
+			coeff_val = coeff_in_bytes[0] +
+				(coeff_in_bytes[1] << 8) +
+				(coeff_in_bytes[2] << 16);
+
+			dai_dmic_write(dmic, base, coeff_val);
+			base += sizeof(uint32_t);
+
+			coeff_in_bytes += 3;
+		}
+	}
+}
+
+/*
+ * @brief Configures the fir coefficients in the PDMs' RAM
+ *
+ * @return Returns pointer right after coefficients data
+ */
+static const uint32_t *dai_dmic_configure_coeff(const struct dai_intel_dmic *dmic,
+						const struct nhlt_pdm_ctrl_cfg * const pdm_cfg,
+						const uint32_t pdm_base,
+						const uint32_t *coeffs)
+{
+	int fir_length_a, fir_length_b;
+	bool packed = false;
+	const uint32_t *coeffs_b;
+
+	fir_length_a = FIELD_GET(FIR_CONFIG_FIR_LENGTH, pdm_cfg->fir_config[0].fir_config) + 1;
+	fir_length_b = FIELD_GET(FIR_CONFIG_FIR_LENGTH, pdm_cfg->fir_config[1].fir_config) + 1;
+
+	if (fir_length_a > 256 || fir_length_b > 256) {
+		LOG_ERR("invalid coeff length! %d %d", fir_length_a, fir_length_b);
+		return NULL;
+	}
+
+	if (*coeffs == FIR_COEFFS_PACKED_TO_24_BITS) {
+		packed = true;
+
+		/* First dword is not included into length_0 and length_1 - skip it. */
+		coeffs++;
+	}
+
+	coeffs_b = dai_dmic_skip_coeff(coeffs, fir_length_a, packed);
+
+	LOG_INF("fir_length_a = %d, fir_length_b = %d, packed = %d", fir_length_a, fir_length_b,
+		packed);
+
+	if (dmic->dai_config_params.dai_index == 0) {
+		dai_dmic_write_coeff(dmic, pdm_base + PDM_COEFFICIENT_A, coeffs, fir_length_a,
+				     packed);
+	} else {
+		dai_dmic_write_coeff(dmic, pdm_base + PDM_COEFFICIENT_B, coeffs_b, fir_length_b,
+				     packed);
+	}
+
+	return dai_dmic_skip_coeff(coeffs_b, fir_length_b, packed);
+}
+
+static int dai_nhlt_get_clock_div(const struct dai_intel_dmic *dmic, const int pdm)
+{
+	uint32_t val;
+	int p_mcic, p_clkdiv, p_mfir, rate_div;
+
+	val = dai_dmic_read(dmic, base[pdm] + CIC_CONFIG);
+	p_mcic = FIELD_GET(CIC_CONFIG_COMB_COUNT, val) + 1;
+
+	val = dai_dmic_read(dmic, base[pdm] + MIC_CONTROL);
+	p_clkdiv = FIELD_GET(MIC_CONTROL_PDM_CLKDIV, val) + 2;
+
+	val = dai_dmic_read(dmic, base[pdm] +
+			    FIR_CHANNEL_REGS_SIZE * dmic->dai_config_params.dai_index + FIR_CONFIG);
+	LOG_ERR("pdm = %d, FIR_CONFIG = 0x%08X", pdm, val);
+
+	p_mfir = FIELD_GET(FIR_CONFIG_FIR_DECIMATION, val) + 1;
+
+	rate_div = p_clkdiv * p_mcic * p_mfir;
+	LOG_ERR("dai_index = %d, rate_div = %d, p_clkdiv = %d, p_mcic = %d, p_mfir = %d",
```

Comment:
@softwarecki why are these `LOG_ERR()`? They seem printed every time:
- https://github.com/thesofproject/sof/issues/8729

This delay https://github.com/thesofproject/sof-test/pull/1075 once more...

Reply:
@marc-hb oops, it should be LOG_INF

Reply:
> ops, it should be LOG_INF

Fix submitted: #67567